newIDE/app/src/ResourcesEditor/ResourcePropertiesEditor/index.js

```diff
@@ -54,9 +65,29 @@ export default class ResourcePropertiesEditor extends React.Component<
     );
   }
 
+  _setResourcePath = () => {
+    const { project, resources, onUpdateProperties } = this.props;
+    const projectPath = path.dirname(project.getProjectFile());
+    const resource = resources[0];
+
+    const options = {
```

Comment:
This won't work in the web-app, because you're calling selectLocalResourcePath, which should be something "hidden".

As I was writing, I think that `onChooseResource` and resources sources (in LocalResourceSources.js and BrowserResourceSources.js) are abstractions that are already there to enable us to show to the user the proper dialog to choose a resource file (either a file on the desktop app, or a url on the web-app). 

Instead of using selectLocalResourcePath, which is something that is "private" the "resource sources" for the desktop app, get the kind of the resource (resource.getKind()) and call the appropriate resource source using `onChooseResource` (read my other previous post, you have to pass it from MainFrame down the this component).
Then get the resource that was returned and use its path:

`onChooseResource(resource.getKind()).then((resources) => {
  if (!resources.length) return; // No path was chosen by the user.

  resource.setFile(resources[0].getPath());
  resources.forEach(resource => resource.delete()); // Important, we are responsible for deleting the resources that were given to us. Otherwise we have a memory leak.

  onUpdateProperties();
  this.forceUpdate();
})`

Reply:
Let me know if this is not clear.
The idea is whenever GDevelop needs a resource path, it should call onChooseResource which will handle showing whatever is needed (for now, a folder selector in the desktop app, but also a dialog in web-app and later it could even be an asset store or something totally different).

Reply:
@4ian the parameter of onChooseResource you have given me is wrong
`onChooseResource(resource.getKind())`
it's not asking for the resource type, but for the platform. To get it to work, I have to do
`onChooseResource('localFileOpener')....`
This poses one problem - the method has no parameter to actually filter the file type to display

Reply:
I can try and add a third optional parameter?

Reply:
you want me to use this function right?
```
  _onChooseResource = (
    sourceName: string,
    multiSelection: boolean = true
  ): Promise<Array<any>>
```

Reply:
we could just put this
`const sources = resourceSources.filter(source => source.kind === fileTypeParameter);`
inside  `_onChooseResource` and then call it like this

`onChooseResource('image',false)`

Reply:
> you want me to use this function right?
> 
> ```
>   _onChooseResource = (
>     sourceName: string,
>     multiSelection: boolean = true
>   ): Promise<Array<any>>
> ```

Yes. Well to be very precise, this is indeed the implementation of the function (but you're right, this is the function that will be ultimately called). For your component you just know that you're getting a prop typed like this: 
`onChooseResource: ChooseResourceFunction,` (with ChooseResourceFunction being typed in [ResourcesList/ResourceSource.flow.js](https://github.com/4ian/GDevelop/blob/2558052a61378fa4edfc9ad53c75d51984eac7e9/newIDE/app/src/ResourcesList/ResourceSource.flow.js#L13-L16))

> the parameter of onChooseResource you have given me is wrong

You're right, the name of the source is depend on the platform.
This is because conceptually, you can have multiple sources for a single type (you could have a file dialog, an asset store, something that generate a random image from the internet, an image picker...). 
For now, let's do as the other editors, it's only two lines:

https://github.com/4ian/GDevelop/blob/2558052a61378fa4edfc9ad53c75d51984eac7e9/newIDE/app/src/ObjectEditor/Editors/SpriteEditor/SpritesList.js#L178-L179

In the future, we may change these 2 lines so that editors ask the user what he want to get the image from (local disk, asset store, etc...) :) But for now, let's keep things simple and stupid and get the first source that match the resource kind:

```
const sources = resourceSources.filter(source => source.kind === resource.getKind());
if (!sources.length) return;

// Pass sources[0].name to onChooseResource
```

> the method seems to be dependent on resourceSources, which needs to also be passed separately everywhere it is needed. Cant we just put the resourceSources logic inside the onChooseResource method, instead of passing it around props and creating more code?

So here you're suffering from the props "drilling", which is the fact of having to pass a lot of props down to React component.
I think that:
- maybe onChooseResource could have been changed to have the resource kind.
- maybe we could have passed a single property containing both the source and the function to choose a resource.
- maybe we could have put this in a React "context", so that we can get them from whenever we want without passing them down in the tree of components (like the user profile).

but it's only two props, it makes things very clear (it's super easy to identify which components are dealing with resources) and even if it's a bit of a pain to pass down properties, it's not hard and once it's done you have good clarity on what your component requires :)

Reply:
This is a bit of work, but this mean that your button will "automagically" be compatible in the future with whatever new source of image/audio/font we create :)

Reply:
my only concern is duplicating code and making it harder to understand, but I think perhaps you have it like this for a good reason. 
All I wanted to add is asimple function that returns a string with the path. With the function there we return resources and get their path to use, then finally we are destroying them to remove them from memory

Reply:
just to be clear you dont want me to add a filetype parameter and move the platform sources logic inside the onchoseresource function? I dont mind either way, just sharing ideas here :)

Reply:
> just to be clear you dont want me to add a filetype parameter and move the platform sources logic inside the onchoseresource function?

No, otherwise when someone will add a new source, like an asset store, components will be able to show it (because they get the resourcesSources) but they won't be able to choose something from it (because onChooseResource would select for you the first source, and will discard all the other sources.). That would not make sense right?

> All I wanted to add is a simple function that returns a string with the path. With the function there we return resources and get their path to use, then finally we are destroying them to remove them from memory

That seems more complicated I admit, but this is because your need is only a specific case of an abstraction that already exist, which is composed of `onChooseResource` and `resourceSources`. This abstraction is very powerful because with only two props, you can display an unlimited number of sources, for any kind of resources, and then open their dialogs or whatever when the user select one, and finally get a resource created for you :) In your case, there is a bit of "waste" because yeah you have to get the path of the resource then delete it.
But honestly, this "waste" has 0% impact on anything.

Reply:
I see, thank you for explaining this. As  said I dont mind it either way, was just wondering why its like this :)

Reply:
No worries. Trying to keep things as modular as possible here because later we could plug many more things, including cloud providers like Dropbox, Google Drive, etc..., which is, given the recent evolution of computing, more and more likely to be asked.